s3upload.js

```diff
@@ -32,16 +37,19 @@ function S3Upload(options) {
             this[option] = options[option];
         }
     }
-    var files = this.fileElement ? this.fileElement.files : this.files || [];
-    this.handleFileSelect(files);
+    this.handleFileSelect(this.files);
 }
 
 S3Upload.prototype.handleFileSelect = function(files) {
     var result = [];
+    var that = this;
+    var next = function (file) {
+        that.onProgress(0, 'Waiting', file);
+        result.push(that.uploadFile(file));
+    }
     for (var i=0; i < files.length; i++) {
         var file = files[i];
-        this.onProgress(0, 'Waiting', file);
-        result.push(this.uploadFile(file));
+        this.onBeforeUpload(file, next)
```

Comment:
Instead of requiring the client pass `file` back to the `next` callback, let's just inline the `next` function here:

```
this.onBeforeUpload(file, (function() {
    this.onProgress(0, 'Waiting', file);
    result.push(this.uploadFile(file));
}).bind(this));